orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py

```diff
@@ -422,3 +465,87 @@ def _try_get_initializer_value(self, model, name):
+    """
+
+    NORMALIZER_FATOR = float(1024 * 1024)
+    NORMALIZER_UNIT = "MiB"
+
+    def __init__(self, m: torch.nn.Module, logger: Logger):
+        self._logger = logger
+        self._current_step = 0
+        self._rank = 0
+        self._world_size = 1
+        if torch.distributed.is_initialized():
+            self._rank = torch.distributed.get_rank()
+            self._world_size = torch.distributed.get_world_size()
+
+        self._rank_info = f"[{self._rank}/{self._world_size}]"
+        self._pre_phase = Phase.INVALID
+        self._last_pahse = Phase.POST_BACKWARD if m.training else Phase.POST_FORWARD
+
+        self._is_first_inspect = True
+
+    def inspect_memory(self, cur_phase: Phase):
+        if not torch.cuda.is_available():
+            return
+
+        if self._is_first_inspect:
+            # Clean the memory cache and memory stats before the first time run forward pass, FOR EVERY RANK.
+            torch.cuda.empty_cache()
+            torch.cuda.reset_peak_memory_stats()
+            self._is_first_inspect = False
+
+        if self._rank != 0:
+            return
+
+        if cur_phase < Phase.PRE_FORWARD or cur_phase > self._last_pahse:
+            raise RuntimeError(f"Invalid phase detected: {cur_phase}")
+
+        if (cur_phase - self._pre_phase) != 1:
+            raise RuntimeError(f"Invalid phase transition detected: {self._pre_phase} -> {cur_phase}")
+
+        cur_mem_allocated = self._normalize(torch.cuda.memory_allocated())
+        max_mem_allocated = self._normalize(torch.cuda.max_memory_allocated())
+        cur_mem_cached = self._normalize(torch.cuda.memory_reserved())
+        max_mem_cached = self._normalize(torch.cuda.max_memory_reserved())
+        torch_mem_stat = torch.cuda.memory_stats()
+        cur_mem_inactive = self._normalize(torch_mem_stat.get("inactive_split_bytes.all.current", 0))
+        max_mem_inactive = self._normalize(torch_mem_stat.get("inactive_split_bytes.all.peak", 0))
+
+        mem_stats = [
+            ["phase", _convert_phase_to_string(cur_phase)],
+            ["allocated", cur_mem_allocated],  # current memory alloeated for tensors
+            ["max allocated", max_mem_allocated],  # peak memory allocated for tensors
+            ["cached", cur_mem_cached],  # current memory cached for caching allocator
+            ["max cached", max_mem_cached],  # peak memory cached for caching allocator.
+            ["inactive", cur_mem_inactive],  # amount of inactive, non-releasable memory
+            ["max inactive", max_mem_inactive],  # peak of inactive, non-releasable memory
+        ]
+
+        summ = f"{self._rank_info} step {self._current_step} memory ({MemoryObserver.NORMALIZER_UNIT})"
+        for stat in mem_stats:
+            summ += f" | {stat[0]}: {stat[1]}"
+
+        # For the 10+ steps, only print when it is power of 2.
+        if self._current_step < 10 or (self._current_step & (self._current_step - 1) == 0):
+            print(summ)
```

Comment:
can we use logger to log this?
something like this: 
```suggestion
            self._logger.info(summ)
```

Reply:
Good catch! Thanks!